File: src/prototype/web/tests/scenario/defense_integration.test.ts
// Integration tests: DefenseIntent flowing into stepSimulation.
// References docs/design/input_system_defense_v1.md §B.5 and
// docs/design/state_machines_v1.md §3.3 / §4.1.

import { describe, expect, it } from "vitest";
import {
  initialActorState,
  initialGameState,
  stepSimulation,
  type GameState,
} from "../../src/state/game_state.js";
import type { InputFrame } from "../../src/input/types.js";
import type { Intent } from "../../src/input/intent.js";
import {
  ZERO_DEFENSE_INTENT,
  type DefenseIntent,
} from "../../src/input/intent_defense.js";
import { INITIAL_ARM_EXTRACTED } from "../../src/state/arm_extracted.js";
import type { HandFSM } from "../../src/state/hand_fsm.js";

function gripped(side: "L" | "R", zone: "SLEEVE_R" | "SLEEVE_L" | "WRIST_R"): HandFSM {
  return Object.freeze({
    side,
    state: "GRIPPED" as const,
    target: zone,
    stateEnteredMs: 0,
    reachDurationMs: 0,
    lastParriedZone: null,
    lastParriedAtMs: Number.NEGATIVE_INFINITY,
  });
}

const NEUTRAL_INTENT: Intent = Object.freeze({
  hip: { hip_angle_target: 0, hip_push: 0, hip_lateral: 0 },
  grip: { l_hand_target: null, l_grip_strength: 0, r_hand_target: null, r_grip_strength: 0 },
  discrete: [],
});

function frame(over: Partial<InputFrame> = {}): InputFrame {
  return Object.freeze({
    timestamp: 0,
    ls: { x: 0, y: 0 },
    rs: { x: 0, y: 0 },
    l_trigger: 0,
    r_trigger: 0,
    buttons: 0,
    button_edges: 0,
    device_kind: "Keyboard" as const,
    ...over,
  });
}

describe("DefenseIntent — posture_break recovery (§3.3 bullet 4)", () => {
  it("weight_forward shoves break toward origin (negative y component)", () => {
    // Seed a pre-existing forward break on the TOP actor; the defender
    // pushes weight forward → sagittal should fall faster than decay alone.
    const seed: GameState = Object.freeze({
      ...initialGameState(0),
      top: Object.freeze({ ...initialActorState(0), postureBreak: { x: 0, y: 0.8 } }),
    });

    // Run 500ms with defender pushing forward.
    const defense: DefenseIntent = Object.freeze({
      hip: { weight_forward: 1, weight_lateral: 0 },
      base: ZERO_DEFENSE_INTENT.base,
      discrete: [],
    });

    let g = seed;
    for (let t = 0; t < 500; t += 16.67) {
      g = stepSimulation(g, frame({ timestamp: t }), NEUTRAL_INTENT, {
        realDtMs: 16.67,
        gameDtMs: 16.67,
        confirmedTechnique: null,
        defenseIntent: defense,
      }).nextState;
    }

    // Compare against pure-decay baseline: same seed, ZERO defense.
    let baseline = seed;
    for (let t = 0; t < 500; t += 16.67) {
      baseline = stepSimulation(baseline, frame({ timestamp: t }), NEUTRAL_INTENT, {
        realDtMs: 16.67,
        gameDtMs: 16.67,
        confirmedTechnique: null,
        defenseIntent: ZERO_DEFENSE_INTENT,
      }).nextState;
    }

    expect(g.top.postureBreak.y).toBeLessThan(baseline.top.postureBreak.y);
  });
});

describe("DefenseIntent — arm_extracted clear via base hold (§4.1)", () => {
  it("RECOVERY_HOLD clears a previously-extracted arm flag", () => {
    // Seed a GameState with arm_extracted already true. Feed pulling
    // conditions so arm_extracted would normally stay; but also feed
    // RECOVERY_HOLD — §4.1 defender-base-hold clears the flag.
    const seed: GameState = Object.freeze({
      ...initialGameState(0),
      bottom: Object.freeze({
        ...initialActorState(0),
        leftHand: gripped("L", "SLEEVE_R"),
      }),
      top: Object.freeze({
        ...initialActorState(0),
        armExtractedLeft: true,
      }),
      topArmExtracted: Object.freeze({
        ...INITIAL_ARM_EXTRACTED,
        right: true,                 // opponent's right arm is extracted
        rightSetAtMs: 0,
      }),
    });

    const defense: DefenseIntent = Object.freeze({
      hip: ZERO_DEFENSE_INTENT.hip,
      base: ZERO_DEFENSE_INTENT.base,
      discrete: [{ kind: "RECOVERY_HOLD" }],
    });

    const res = stepSimulation(seed, frame({
      timestamp: 16,
      l_trigger: 0.9, // keep pulling
    }), {
      ...NEUTRAL_INTENT,
      hip: { hip_angle_target: 0, hip_push: -0.6, hip_lateral: 0 },
      grip: {
        l_hand_target: "SLEEVE_R",
        l_grip_strength: 0.9,
        r_hand_target: null,
        r_grip_strength: 0,
      },
    }, {
      realDtMs: 16,
      gameDtMs: 16,
      confirmedTechnique: null,
      defenseIntent: defense,
    });

    expect(res.nextState.topArmExtracted.right).toBe(false);
    expect(res.nextState.top.armExtractedRight).toBe(false);
  });
});

File: src/prototype/unity/Assets/BJJSimulator/Tests/EditMode/Scenario/DefenseIntegrationTest.cs
// NUnit EditMode mirror of src/prototype/web/tests/scenario/defense_integration.test.ts.
// Each [Test] corresponds to one it(...) case from the Stage 1 Vitest suite so a
// regression on either side produces a named, greppable failure.
//
// Integration tests: DefenseIntent flowing into stepSimulation.
// References docs/design/input_system_defense_v1.md §B.5 and
// docs/design/state_machines_v1.md §3.3 / §4.1.
//
// Run from Unity Editor: Window → General → Test Runner → EditMode.

using NUnit.Framework;

namespace BJJSimulator.Tests.Scenario
{
    // -------------------------------------------------------------------------
    // Helpers — mirror the gripped() / frame() factories and NEUTRAL_INTENT in TS.
    // -------------------------------------------------------------------------

    public static class DefenseIntegrationHelpers
    {
        public static HandFSM Gripped(HandSide side, GripZone zone) => new HandFSM
        {
            Side            = side,
            State           = HandState.Gripped,
            Target          = zone,
            StateEnteredMs  = 0,
            ReachDurationMs = 0,
            LastParriedZone = GripZone.None,
            LastParriedAtMs = BJJConst.SentinelTimeMs,
        };

        public static readonly Intent NeutralIntent = new Intent
        {
            Hip      = HipIntent.Zero,
            Grip     = GripIntent.Zero,
            Discrete = System.Array.Empty<DiscreteIntent>(),
        };

        public static InputFrame Frame(
            long timestamp = 0,
            float lTrigger = 0f,
            float rTrigger = 0f) =>
            new InputFrame
            {
                Timestamp   = timestamp,
                Ls          = Vec2.Zero,
                Rs          = Vec2.Zero,
                LTrigger    = lTrigger,
                RTrigger    = rTrigger,
                Buttons     = ButtonBit.None,
                ButtonEdges = ButtonBit.None,
                DeviceKind  = DeviceKind.Keyboard,
            };

        public static StepOptions Opts(
            float dtMs = 16.67f,
            Technique? confirmedTechnique = null,
            DefenseIntent? defense = null) =>
            new StepOptions
            {
                RealDtMs           = dtMs,
                GameDtMs           = dtMs,
                ConfirmedTechnique = confirmedTechnique,
                ConfirmedCounter   = null,
                DefenseIntent      = defense,
            };
    }

    // -------------------------------------------------------------------------
    // describe("DefenseIntent — posture_break recovery (§3.3 bullet 4)")
    // -------------------------------------------------------------------------

    [TestFixture]
    public class DefenseIntentPostureBreakRecoveryTests
    {
        // it("weight_forward shoves break toward origin (negative y component)")
        [Test]
        public void WeightForwardShovesBreakTowardOriginNegativeYComponent()
        {
            // Seed a pre-existing forward break on the TOP actor; the defender
            // pushes weight forward → sagittal should fall faster than decay alone.
            var seed = GameStateOps.InitialGameState(0);
            var top = seed.Top;
            top.PostureBreak = new Vec2(0f, 0.8f);
            seed.Top = top;

            // Run 500ms with defender pushing forward.
            var defense = new DefenseIntent
            {
                Hip      = new TopHipIntent { WeightForward = 1f, WeightLateral = 0f },
                Base     = DefenseIntent.Zero.Base,
                Discrete = System.Array.Empty<DefenseDiscreteIntent>(),
            };

            var g = seed;
            for (float t = 0; t < 500f; t += 16.67f)
            {
                g = GameStateOps.Step(
                    g,
                    DefenseIntegrationHelpers.Frame(timestamp: (long)t),
                    DefenseIntegrationHelpers.NeutralIntent,
                    DefenseIntegrationHelpers.Opts(
                        dtMs:    16.67f,
                        defense: defense)).NextState;
            }

            // Compare against pure-decay baseline: same seed, ZERO defense.
            var baseline = seed;
            for (float t = 0; t < 500f; t += 16.67f)
            {
                baseline = GameStateOps.Step(
                    baseline,
                    DefenseIntegrationHelpers.Frame(timestamp: (long)t),
                    DefenseIntegrationHelpers.NeutralIntent,
                    DefenseIntegrationHelpers.Opts(
                        dtMs:    16.67f,
                        defense: DefenseIntent.Zero)).NextState;
            }

            Assert.Less(g.Top.PostureBreak.Y, baseline.Top.PostureBreak.Y);
        }
    }

    // -------------------------------------------------------------------------
    // describe("DefenseIntent — arm_extracted clear via base hold (§4.1)")
    // -------------------------------------------------------------------------

    [TestFixture]
    public class DefenseIntentArmExtractedClearTests
    {
        // it("RECOVERY_HOLD clears a previously-extracted arm flag")
        [Test]
        public void RecoveryHoldClearsAPreviouslyExtractedArmFlag()
        {
            // Seed a GameState with arm_extracted already true. Feed pulling
            // conditions so arm_extracted would normally stay; but also feed
            // RECOVERY_HOLD — §4.1 defender-base-hold clears the flag.
            var seed = GameStateOps.InitialGameState(0);
            var bottom = seed.Bottom;
            bottom.LeftHand = DefenseIntegrationHelpers.Gripped(HandSide.L, GripZone.SleeveR);
            seed.Bottom = bottom;
            var top = seed.Top;
            top.ArmExtractedLeft = true;
            seed.Top = top;
            seed.TopArmExtracted = new ArmExtractedState
            {
                Left           = ArmExtractedState.Initial.Left,
                Right          = true,                       // opponent's right arm is extracted
                LeftSustainMs  = ArmExtractedState.Initial.LeftSustainMs,
                RightSustainMs = ArmExtractedState.Initial.RightSustainMs,
                LeftSetAtMs    = ArmExtractedState.Initial.LeftSetAtMs,
                RightSetAtMs   = 0,
            };

            var defense = new DefenseIntent
            {
                Hip      = DefenseIntent.Zero.Hip,
                Base     = DefenseIntent.Zero.Base,
                Discrete = new[] { new DefenseDiscreteIntent { Kind = DefenseDiscreteIntentKind.RecoveryHold } },
            };

            var pullingIntent = new Intent
            {
                Hip = new HipIntent { HipAngleTarget = 0f, HipPush = -0.6f, HipLateral = 0f },
                Grip = new GripIntent
                {
                    LHandTarget   = GripZone.SleeveR,
                    LGripStrength = 0.9f,
                    RHandTarget   = GripZone.None,
                    RGripStrength = 0f,
                },
                Discrete = System.Array.Empty<DiscreteIntent>(),
            };

            var (next, _) = GameStateOps.Step(
                seed,
                DefenseIntegrationHelpers.Frame(timestamp: 16, lTrigger: 0.9f), // keep pulling
                pullingIntent,
                DefenseIntegrationHelpers.Opts(
                    dtMs:    16f,
                    defense: defense));

            Assert.IsFalse(next.TopArmExtracted.Right);
            Assert.IsFalse(next.Top.ArmExtractedRight);
        }
    }
}
